backend/internal/sessionsgrpcsvc/svc.go

```diff
@@ -125,27 +124,55 @@ func (s *server) List(ctx context.Context, req *connect.Request[sessionsv1.ListR
 	var err error
 
 	if filter.DeploymentID, err = sdktypes.ParseDeploymentID(req.Msg.DeploymentId); err != nil {
-		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("deployment_id: %w", err))
+		return nil, wrapError(fmt.Errorf("deployment_id: %w", err), connect.CodeInvalidArgument)
 	}
 
 	if filter.EventID, err = sdktypes.ParseEventID(req.Msg.EventId); err != nil {
-		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("event_id: %w", err))
+		return nil, wrapError(fmt.Errorf("event_id: %w", err), connect.CodeInvalidArgument)
 	}
 
 	if filter.EnvID, err = sdktypes.ParseEnvID(req.Msg.EnvId); err != nil {
-		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("env_id: %w", err))
+		return nil, wrapError(fmt.Errorf("env_id: %w", err), connect.CodeInvalidArgument)
 	}
 
 	sessions, n, err := s.sessions.List(ctx, filter)
 	if err != nil {
-		if errors.Is(err, sdkerrors.ErrUnauthorized) {
-			return nil, connect.NewError(connect.CodePermissionDenied, err)
-		}
-
-		return nil, connect.NewError(connect.CodeUnknown, err)
+		return nil, wrapError(err, 0)
 	}
 
 	pbsessions := kittehs.Transform(sessions, sdktypes.ToProto)
 
 	return connect.NewResponse(&sessionsv1.ListResponse{Sessions: pbsessions, Count: int32(n)}), nil
 }
+
+func (s *server) Delete(ctx context.Context, req *connect.Request[sessionsv1.DeleteRequest]) (*connect.Response[sessionsv1.DeleteResponse], error) {
+	msg := req.Msg
+
+	if err := proto.Validate(msg); err != nil {
+		return nil, wrapError(err, connect.CodeInvalidArgument)
+	}
+
+	sessionID, err := sdktypes.ParseSessionID(msg.SessionId)
+	if err != nil {
+		return nil, wrapError(fmt.Errorf("session_id: %w", err), connect.CodeInvalidArgument)
+	}
+
+	session, err := s.sessions.Get(ctx, sessionID)
+	if err != nil {
+		return nil, wrapError(err, 0)
+	}
+
+	if session == nil { // just to ensure, Get should return an error if not found
+		return nil, wrapError(err, connect.CodeNotFound)
+	}
+
+	// FIXME: maybe connect.CodeCancelled?
+	if state := sdktypes.GetSessionLatestState(session); state == sdktypes.RunningSessionStateType {
```

Comment:
created is also a problematic state
I think you should check if state is not completed or failed

Reply:
rewritten